### PR TITLE
Added backwards compatible way to do sanity checking on permissions and raise exceptions when you make a typo.

### PR DESCRIPTION
**Pull request by hvdklauw**

### CODE REVIEW

File: balrog/policy.py [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
 from balrog import exceptions
+from balrog.permission import Permission
 
 
 class Policy(object):
 
     """Controls the access of a certain actor to a certain action on a resource."""
 
-    def __init__(self, roles, get_identity, get_role):
+    def __init__(self, roles, get_identity, get_role, permissions=None):
```
Comment (bubenkoff) [inline balrog/policy.py]: policy is able to get all possible permissions by itself
there's no need to receive it as a parameter

Comment (hvdklauw) [inline balrog/policy.py]: Yes, but this way you can NOT do the checking by not providing it.

We could get them from the roles but we'll start raising exceptions while
before we didn't.

Comment (bubenkoff) [inline balrog/policy.py]: getting them from the roles is exactly what i mean

On 26 November 2014 at 14:00, Harro van der Klauw notifications@github.com
wrote:

> In balrog/policy.py:
> 
> >  class Policy(object):
> > 
> > ```
> >  """Controls the access of a certain actor to a certain action on a resource."""
> > ```
> > -    def **init**(self, roles, get_identity, get_role):
> > -    def **init**(self, roles, get_identity, get_role, permissions=None):
> 
> Yes, but this way you can NOT do the checking by not providing it. We
> could get them from the roles but we'll start raising exceptions while
> before we didn't.
> … <#149ec2ffe08c3170_>
> On 26 November 2014 at 13:57, Anatoly Bubenkov notifications@github.com
> wrote: In balrog/policy.py: > > > class Policy(object): > > """Controls the
> access of a certain actor to a certain action on a resource.""" > > - def
> **init**(self, roles, get_identity, get_role): > + def **init**(self,
> roles, get_identity, get_role, permissions=None): policy is able to get all
> possible permissions by itself there's no need to receive it as a parameter
> — Reply to this email directly or view it on GitHub <
> https://github.com/paylogic/balrog/pull/1/files#r20933058>.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/paylogic/balrog/pull/1/files#r20933170.

## 

Anatoly Bubenkov

**Comment by hvdklauw** [comment]
One big downside of this: If you have something implemented and all the permission checking in place and then want to quickly disable the feature you need to actually have a single role still have those permissions or your whole system will raise exceptions.

**Comment by hvdklauw** [comment]
On the other hand we can just say: If you use it and don't want the exceptions handle it in your wrapper/decorator together with the "False" result.

File: balrog/policy.py [hunk 2]
```diff
@@ -51,9 +67,14 @@ def check(self, permission, *args, **kwargs):
...
         :param permission: Permission name.
         :return: `True` if identity role has this permission.
+        :raises: `PermissionNotFound` when no permission is found.
         """
+        self._check_permission(permission)
         identity = self.get_identity(*args, **kwargs)
-        role = self.get_role(identity, *args, **kwargs)
+        try:
+            role = self.get_role(identity, *args, **kwargs)
+        except exceptions.RoleNotFound:
```
Comment (bubenkoff) [inline balrog/policy.py]: what is the reason for this change?

Comment (bubenkoff) [inline balrog/policy.py]: we loose the check for invalid roles, but we add a check for invalid persmissions, well, in total it's 0 :)

Comment (hvdklauw) [inline balrog/policy.py]: good point, I'll revert

**Comment by hvdklauw** [comment]
@bubenkoff fixed :)
